File: cpongo/token.c
//
// Created by Tre on 5/31/24.
//

#include "token.h"
#include "common.h"
#include "string.h"

void print_token(Token t) {
    char type[81] = {0};

    switch (t.type) {
        case LEFT_PAREN:
            strcpy(type, "LEFT_PAREN");
            break;
        case RIGHT_PAREN:
            strcpy(type, "RIGHT_PAREN");
            break;
        case QUOTE:
            strcpy(type, "QUOTE");
            break;
        case SEMICOLON:
            strcpy(type, "SEMICOLON");
            break;
        case PLUS:
            strcpy(type, "PLUS");
            break;
        case MINUS:
            strcpy(type, "MINUS");
            break;
        case STAR:
            strcpy(type, "STAR");
            break;
        case SLASH:
            strcpy(type, "SLASH");
            break;
        case PERCENT:
            strcpy(type, "PERCENT");
            break;
        case AND:
            strcpy(type, "AND");
            break;
        case OR:
            strcpy(type, "OR");
            break;
        case NOT:
            strcpy(type, "NOT");
            break;
        case ASSIGN:
            strcpy(type, "ASSIGN");
            break;
        case SHORT:
            strcpy(type, "SHORT");
            break;
        case EQUAL:
            strcpy(type, "EQUAL");
            break;
        case NOT_EQUAL:
            strcpy(type, "NOT_EQUAL");
            break;
        case GREATER:
            strcpy(type, "GREATER");
            break;
        case GREATER_EQUAL:
            strcpy(type, "GREATER_EQUAL");
            break;
        case LESS:
            strcpy(type, "LESS");
            break;
        case LESS_EQUAL:
            strcpy(type, "LESS_EQUAL");
            break;
        case IDENTIFIER:
            strcpy(type, "IDENTIFIER");
            break;
        case STRING:
            strcpy(type, "STRING");
            break;
        case DECIMAL:
            strcpy(type, "DECIMAL");
            break;
        case HEX:
            strcpy(type, "HEX");
            break;
        case TRUE:
            strcpy(type, "TRUE");
            break;
        case FALSE:
            strcpy(type, "FALSE");
            break;
        case RAND:
            strcpy(type, "RAND");
            break;
        case EXIT:
            strcpy(type, "EXIT");
            break;
        case SMASH:
            strcpy(type, "SMASH");
            break;
        case SMASHALL:
            strcpy(type, "SMASHALL");
            break;
        case INPUT:
            strcpy(type, "INPUT");
            break;
        case PRINT:
            strcpy(type, "PRINT");
            break;
        case PRINTLN:
            strcpy(type, "PRINTLN");
            break;
        case CLOCK:
            strcpy(type, "CLOCK");
            break;
        case GOTO:
            strcpy(type, "GOTO");
            break;
        case THENGOTO:
            strcpy(type, "THENGOTO");
            break;
        case LABEL:
            strcpy(type, "LABEL");
            break;
        case END_OF_FILE:
            strcpy(type, "END_OF_FILE");
            break;
        default: // this shouldnt happen
            strcpy(type, "UNRECOGNIZED");
            break;
    }

    printf("%s \"%s\", line %d\n", type, t.lexeme, t.line);
}

File: token/token.go
package token

const (
	UNKNOWN = "UNKNOWN" // for unrecognized tokens
	ERROR = "ERROR" // returned if lexer encounters error parsing another token
	EOF = "EOF"

	// identifiers and literals
	IDENT = "IDENT"
	DECIMAL = "DECIMAL" // base-10 number literal
	HEX = "HEX" // base-16 number literal
	STRINGLIT = "STRINGLIT" // string literal
	
	// operators
	ASSIGN = "="
	PLUS = "+"
	MINUS = "-"
	TIMES = "*"
	DIVIDE = "/"
	MODULO = "%"
	BANG = "!"
	AND = "&"
	OR = "|"
  AT = "@" // like a "get" operator

  // comparison operators
	LT = "<"
	GT = ">"
	LE = "<="
	GE = ">="
	EQ = "=="
	NOT_EQ = "!="

	// delims
	SEMICOLON = ";"
	COMMA = ","
	LPAREN = "("
	RPAREN = ")"

	// keywords
	SHORT = "SHORT"
	SMASH = "SMASH"
	SMASHALL = "SMASHALL"
	GOTO = "GOTO"
	// GOBACK = "GOBACK"
	IF = "IF"
	EXIT = "EXIT"
	PRINT = "PRINT"
	PRINTLN = "PRINTLN"
	CLOCK = "CLOCK"
	INPUT = "INPUT"
	TRUE = "TRUE"
	FALSE = "FALSE"
	RAND = "RAND"
	LBL = "LBL"
  BUFF = "BUFF"
  SIZEOF = "SIZEOF"
  // OP = "OP"
  // NOTE: would I rather do pure-functions over ops?
)

type TokenType string

type Token struct {
	Type TokenType
	Literal string
	LineNumber int // store line number in token for error messages
}

var keywords = map[string]TokenType {
  "short": SHORT,
  "smash": SMASH,
  "smashall": SMASHALL,
  "goto": GOTO,
  "if": IF,
  // "goback": GOBACK,
  "exit": EXIT,
  "print": PRINT,
  "println": PRINTLN,
  "clock": CLOCK,
  "input": INPUT,
  "true": TRUE,
  "false": FALSE,
  "rand": RAND,
  "lbl": LBL,
  "buff": BUFF,
  "sizeof": SIZEOF,
  // "op": OP,
}

func LookupIdent(ident string) TokenType {
  if tok, ok := keywords[ident]; ok {
    return tok
  }
  return IDENT
}